crates/rome_js_formatter/src/ts/types/union_type.rs

```diff
@@ -1,67 +1,110 @@
+                parent_kind => {
+                    matches!(
+                        parent_kind,
+                        Some(
+                            JsSyntaxKind::TS_TYPE_ASSERTION_EXPRESSION
+                                | JsSyntaxKind::TS_TUPLE_TYPE_ELEMENT_LIST
+                                | JsSyntaxKind::TS_TYPE_ASSERTION_ASSIGNMENT
+                                | JsSyntaxKind::TS_TYPE_ARGUMENT_LIST
+                        )
+                    )
+                }
+            }
+        };
+
+        let types = format_with(|f| {
             write!(
                 f,
                 [
-                    soft_line_break(),
                     FormatTypeSetLeadingSeparator {
                         separator: JsSyntaxKind::PIPE,
-                        leading_separator: leading_separator_token.as_ref()
+                        leading_separator: leading_separator_token.as_ref(),
+                        leading_soft_line_break_or_space: should_indent
+                            && !has_leading_own_line_comment,
                     },
                     types.format()
                 ]
             )
         });
 
-        if node.needs_parentheses() {
-            return write!(f, [group(&format_args![indent(&body), soft_line_break()])]);
-        }
+        let content = format_with(|f| {
+            // it is necessary to add parentheses for unions in intersections
+            // ```ts
+            // type Some = B & (C | A) & D
+            // ```
+            if node.needs_parentheses() {
+                return write!(f, [indent(&types), soft_line_break()]);
+            }
 
-        let should_indent = {
-            let parent_kind = node.syntax().parent().map(|p| p.kind());
+            let is_inside_complex_tuple_type = node
+                .parent::<TsTupleTypeElementList>()
+                .map_or(false, |tuple| tuple.len() > 1);
 
-            !matches!(
-                parent_kind,
-                Some(
-                    JsSyntaxKind::TS_REFERENCE_TYPE
-                        | JsSyntaxKind::TS_TYPE_ASSERTION_EXPRESSION
-                        | JsSyntaxKind::TS_TUPLE_TYPE
-                        | JsSyntaxKind::TS_TYPE_ASSERTION_ASSIGNMENT
-                        | JsSyntaxKind::TS_FUNCTION_TYPE
-                        | JsSyntaxKind::TS_TYPE_ARGUMENTS
+            if is_inside_complex_tuple_type {
+                write!(
+                    f,
+                    [
+                        indent(&format_args![
+                            if_group_breaks(&format_args![text("("), soft_line_break()]),
+                            types
+                        ]),
+                        soft_line_break(),
+                        if_group_breaks(&text(")"))
+                    ]
                 )
-            )
-        };
+            } else if should_indent {
+                write!(f, [&indent(&types)])
+            } else {
+                write!(f, [&types])
```

Comment:
Nit: You don't need to pass references to `write` and `format_args`
```suggestion
                write!(f, [indent(&types)])
            } else {
                write!(f, [types])
```

Reply:
fixed (=